### ISSUE TITLE
Form input helpers don't generate well-formatted JSON - Upgrade from 4.2.0 to 4.2.8

### ISSUE THREAD

**Issue by acanyon** [rails/rails#28591]
Form input helpers don't handle JSON results.

### Steps to reproduce

```
class Item < ActiveRecord::Base
  serialize :tags, JSON
end

# template
= form_for [:admin, @item]
  .field
    = f.label :tags
    = f.text_field :tags
```
The input field is populated with `midi` instead of `["midi"]`. And similarly in the case of dictionaries, it will return `=>` instead of proper JSON.

### Expected behavior
Should return well-formatted JSON in the input field.

### Actual behavior
Returns string-spaced elements and `=>` for JSON dicts

### System configuration
**Ruby version**:
Upgraded to `4.2.8` from `4.2.0`.

**Comment by pixeltrix** [comment]
@acanyon what the form helpers are seeing isn't JSON, but Ruby data structures like Array and Hash - a straightforward JSON string would be fine, e.g.:

```
>> helper.text_field :foo, :bar, value: ['baz']
=> "<input value=\"baz\" type=\"text\" name=\"foo[bar]\" id=\"foo_bar\" />"
>> helper.text_field :foo, :bar, value: "['baz']"
=> "<input value=\"[&#39;baz&#39;]\" type=\"text\" name=\"foo[bar]\" id=\"foo_bar\" />"
```

The form helpers don't do any introspection of column types so the best way to handle this is to create some accessors that will convert from/to JSON for you and refer to those in your form, e.g:

``` ruby
class Item < ActiveRecord::Base
  serialize :tags, JSON

  def raw_tags
    tags.to_json
  end

  def raw_tags=(json)
    self.tags = JSON.parse(json)
  end
end
```
``` erb
<%= form_for @item do |f| %>
<%= f.text_field :raw_tags %>
<%= end %>
```

Obviously you'd want to put some error checking in there but that gives you a starting point.

**Comment by acanyon** [comment]
Shouldn't the text stored in the DB be returned then? Obviously a depreciation.

**Comment by acanyon** [comment]
Shouldn't this return whats stored in the DB? Surely a depreciation for a minor version.

**Comment by pixeltrix** [comment]
> Shouldn't this return whats stored in the DB? Surely a depreciation for a minor version.

Not every form field has a database column - it's what the attribute returns.

Digging a bit deeper shows that #18523 registered the behaviour you desire as a regression bug which is why it was fixed in 787e22b. However that commit has since been reverted in 5f63fef because we use `*_before_type_cast` in case there's an error in the form. Along with that change was another in 3e74db6 which added a check for whether the value came from the user so that #18523 didn't become an issue again.

Therefore it's working as expected and changed between 4.2.0 and 4.2.8 because of a regression from 4.0 and 4.1.

**Comment by acanyon** [comment]
Sounds solid to me. Thanks for looking into it!
